app/models/service_template.rb

```diff
@@ -538,4 +526,22 @@ def resource_actions_info
   def generic_custom_buttons
     CustomButton.buttons_for("Service")
   end
+
+  def adjust_service_type
+    return if service_type == self.class::SERVICE_TYPE_INTERNAL # internal cannot turn to composite
+
+    if service_resources.blank?
+      svc_type = self.class.SERVICE_TYPE_UNKNOWN
+    else
+      svc_type = self.class::SERVICE_TYPE_ATOMIC
+      service_resources.each do |sr|
```

Comment:
Wouldn't this be more efficient as:
```ruby
service_resources.where(:resource_type => ["Service", "ServiceTemplate"]).any? ? self.class::SERVICE_TYPE_COMPOSITE : self.class::SERVICE_TYPE_ATOMIC
```

Reply:
I like your proposed way. However it does not pass the spec test. I realize this method is called after `add_resource` which is not yet persisted in DB. This method is renamed from `set_service_type`. Let's keep the old implementation which make no assumption that `service_resources` have to be loaded from DB.